examples/raytracing/example_raytracing.cpp

```diff
@@ -0,0 +1,232 @@
+{
+  Kokkos::View<ArborX::Sphere *, MemorySpace> _spheres;
+};
+
+template <typename MemorySpace>
+struct Predicates
+{
+  Kokkos::View<ArborX::Ray *, MemorySpace> _rays;
+};
+
+template <typename MemorySpace>
+struct ArborX::AccessTraits<Primitives<MemorySpace>, ArborX::PrimitivesTag>
+{
+  KOKKOS_FUNCTION static std::size_t
+  size(const Primitives<MemorySpace> &primitives)
+  {
+    return primitives._spheres.extent(0);
+  }
+  KOKKOS_FUNCTION static ArborX::Box
+  get(Primitives<MemorySpace> const &primitives, std::size_t i)
+  {
+    auto const &sphere = primitives._spheres(i);
+    auto const &c = sphere.centroid();
+    auto const &r = sphere.radius();
+    return {{c[0] - r, c[1] - r, c[2] - r}, {c[0] + r, c[1] + r, c[2] + r}};
+  }
+  using memory_space = MemorySpace;
+};
+
+template <typename MemorySpace>
+struct ArborX::AccessTraits<Predicates<MemorySpace>, ArborX::PredicatesTag>
+{
+  KOKKOS_FUNCTION static std::size_t
+  size(const Predicates<MemorySpace> &predicates)
+  {
+    return predicates._rays.extent(0);
+  }
+  KOKKOS_FUNCTION static auto get(Predicates<MemorySpace> const &predicates,
+                                  std::size_t i)
+  {
+    return attach(intersects(predicates._rays(i)), (int)i);
+  }
+  using memory_space = MemorySpace;
+};
+
+template <typename MemorySpace>
+struct Callback
+{
+  using Vector = ArborX::Point;
+
+  Kokkos::View<ArborX::Sphere *, MemorySpace> spheres;
+
+  Kokkos::View<float *, MemorySpace, Kokkos::MemoryTraits<Kokkos::Atomic>>
+      accumulator;
+
+  KOKKOS_FUNCTION float dotproduct(Vector const &v1, Vector const &v2) const
+  {
+    return v1[0] * v2[0] + v1[1] * v2[1] + v1[2] * v2[2];
+  }
+
+  KOKKOS_FUNCTION float solveQuadratic(const float &b, const float &c) const
+  {
+    float delta = b * b - 4.0 * c;
+    if (delta <= 0.0)
+      // NO intersection
+      return 0.0;
+    else
+    {
+      float q = (b > 0) ? -0.5 * (b + std::sqrt(delta))
+                        : -0.5 * (b - std::sqrt(delta));
+
+      // printf("q=%f, c/q = %f \n", q, c / q);
+
+      if (q < 0)
+      {
+        // both intersections are BEHIND
+        if (c > 0)
+          return 0.0;
+        // one point is behind
+        else
+          return c / q;
+      }
+      else
+      {
+        // both intersections are AHEAD
+        if (c > 0)
+          return std::abs(q - c / q);
+        // one point is behind
+        else
+          return q;
+      }
+    }
+  }
+
+  KOKKOS_FUNCTION float overlap_distance(ArborX::Ray const &ray,
+                                         ArborX::Sphere const &sphere) const
+  {
+    auto center = sphere.centroid();
+    auto r = sphere.radius();
+
+    ArborX::Point orig = ray.origin();
+
+    Vector L{orig[0] - center[0], orig[1] - center[1], orig[2] - center[2]};
+
+    //  for normalized direction vector a = 1.0
+    auto dir = ray.direction();
+
+    float b = 2.0 * dotproduct(dir, L);
+    float c = dotproduct(L, L) - r * r;
+
+    return solveQuadratic(b, c);
```

Comment:
Please document an algorithm you are using, explaining how you calculate the length.

Reply:
OK, here's how I would document it:
```
Finding the length of the ray intersection with a sphere consists of two steps. First, find two points on a sphere that a bidirectional ray intersects with. These points are the solution to the equation:
   || (ray_origin + alpha * direction) - sphere_center || = radius.
The resulting quadratic equation is
   || direction ||^2 alpha^2 + 2*(ray_origin - sphere_center, direction) alpha + (radius^2 - ||ray_origin - sphere_center||^2) = 0.

If this equation has a solution, ray intersects with a sphere. Once the alpha values are computed, say alpha_1 and alpha_2 (which could be the same if ray just touches the sphere), the directional ray intersection part will be the interval [alpha_1, alpha_2] intersected with [0, +inf). If intersection is empty, directional ray does not intersect. Once that interval is computed, say, [a, b], the length of intersection is simply (b-a)*||direction||.
```